vcd/metadata_openapi.go

```diff
@@ -0,0 +1,325 @@
+	metadataToAdd, metadataToUpdate, metadataToDelete, err := getOpenApiMetadataOperations(oldRaw.(*schema.Set).List(), newRaw.(*schema.Set).List())
+	if err != nil {
+		return fmt.Errorf("could not calculate the needed metadata operations: %s", err)
+	}
+
+	for _, entry := range metadataToDelete {
+		err = resource.DeleteMetadata(entry.KeyValue.Namespace, entry.KeyValue.Key)
+		if err != nil {
+			return fmt.Errorf("error deleting metadata with namespace '%s' and key '%s': %s", entry.KeyValue.Namespace, entry.KeyValue.Key, err)
+		}
+	}
+
+	for _, entry := range metadataToUpdate {
+		_, err = resource.UpdateMetadata(entry.KeyValue.Namespace, entry.KeyValue.Key, entry.KeyValue.Value.Value)
+		if err != nil {
+			return fmt.Errorf("error updating metadata with namespace '%s' and key '%s': %s", entry.KeyValue.Namespace, entry.KeyValue.Key, err)
+		}
+	}
+
+	for _, metadataEntry := range metadataToAdd {
+		_, err := resource.AddMetadata(metadataEntry)
+		if err != nil {
+			return fmt.Errorf("error adding metadata entry: %s", err)
+		}
+	}
+	return nil
+}
+
+// getOpenApiMetadataOperations retrieves the metadata that needs to be added, to be updated and to be deleted depending
+// on the old and new attribute values from Terraform state.
+func getOpenApiMetadataOperations(oldMetadata []interface{}, newMetadata []interface{}) ([]types.OpenApiMetadataEntry, []types.OpenApiMetadataEntry, []types.OpenApiMetadataEntry, error) {
+	oldMetadataEntries, err := getOpenApiMetadataEntryMap(oldMetadata)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	newMetadataEntries, err := getOpenApiMetadataEntryMap(newMetadata)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var metadataToRemove []types.OpenApiMetadataEntry
+	for oldNamespacedKey := range oldMetadataEntries {
+		if _, ok := newMetadataEntries[oldNamespacedKey]; !ok {
+			metadataToRemove = append(metadataToRemove, oldMetadataEntries[oldNamespacedKey])
+		}
+	}
+
+	var metadataToCreate []types.OpenApiMetadataEntry
+	metadataToUpdateMap := map[string]types.OpenApiMetadataEntry{}
+	for newNamespacedKey, newEntry := range newMetadataEntries {
+		if oldEntry, ok := oldMetadataEntries[newNamespacedKey]; ok {
+			if reflect.DeepEqual(oldEntry, newEntry) {
+				continue
+			}
+			if oldEntry.IsReadOnly != newEntry.IsReadOnly || oldEntry.IsPersistent != newEntry.IsPersistent ||
+				oldEntry.KeyValue.Namespace != newEntry.KeyValue.Namespace || oldEntry.KeyValue.Domain != newEntry.KeyValue.Domain ||
+				oldEntry.KeyValue.Value.Type != newEntry.KeyValue.Value.Type {
+				util.Logger.Printf("[DEBUG] entry with namespace '%s' and key '%s' is being deleted and re-created", oldEntry.KeyValue.Namespace, oldEntry.KeyValue.Key)
+				metadataToRemove = append(metadataToRemove, oldMetadataEntries[newNamespacedKey])
+				metadataToCreate = append(metadataToCreate, newMetadataEntries[newNamespacedKey])
+			} else {
+				metadataToUpdateMap[newNamespacedKey] = newEntry
+			}
```

Comment:
So the last hook here for me to clarify (and possibly worth a comment in this block) - update can happen only if value `KeyValue.Value.Value` change? Otherwise we remove/create metadata, right? I think you put it in website docs, but sucha comment explaining this code block could help is in future.

Reply:
Good point, added a couple of comments.